xdsl/transforms/function_specialization.py

```diff
@@ -0,0 +1,222 @@
+from typing import Iterable, cast
+
+from xdsl.dialects import arith, builtin, func, scf
+from xdsl.dialects.builtin import ArrayAttr, StringAttr
+from xdsl.ir import Attribute, Block, MLContext, Operation, Region
+from xdsl.passes import ModulePass
+from xdsl.pattern_rewriter import (
+    PatternRewriter,
+    PatternRewriteWalker,
+    RewritePattern,
+    op_type_rewrite_pattern,
+)
+from xdsl.traits import SymbolTable
+
+SPEZIALIZE_ON_VALS_ATTR = "specialize_on_vals"
+
+
+class FunctionSpecializationPattern(RewritePattern):
+    @op_type_rewrite_pattern
+    def match_and_rewrite(self, func_op: func.FuncOp, rewriter: PatternRewriter, /):
+        # check if the function contains a "specialize_on_vals" annotated operation
+        split_op = func_contains_specialization_annotation(func_op)
+        if split_op is None:
+            return
+
+        # can't do splits on multi-value ops
+        if len(split_op.results) != 1:
+            return
+
+        # get list of vals to specialize on:
+        specialization_vals = get_op_specialization(split_op)
+        # drop malformed or empty specializations:
+        if specialization_vals is None or len(specialization_vals) == 0:
+            split_op.attributes.pop(SPEZIALIZE_ON_VALS_ATTR)
+            return
+
+        return_types = tuple(func_op.function_type.outputs)
+
+        function_remainder = split_op.next_op
+        assert function_remainder is not None, "Can't specialize on terminator!"
```

Comment:
Can you avoid making the pass fail and maybe bail out instead?

Reply:
Done!